### PR TITLE
Add non-standard packed mode

### LINKED ISSUE #14: Add non-standard packed encoding
> Through abi.encodePacked(), Solidity supports a non-standard packed mode where:
> - types shorter than 32 bytes are concatenated directly, without padding or sign extension
> - dynamic types are encoded in-place and without the length.
> - array elements are padded, but still encoded in-place

—[Contract ABI specification](https://docs.soliditylang.org/en/develop/abi-spec.html#non-standard-packed-mode)

We should support this in `abi-utils`.

### PR DESCRIPTION
**Pull request by Mrtenz**
Closes #14.

This adds the non-standard packed mode, which behaves the same as `abi.encodePacked` in Solidity.

### CODE REVIEW

File: src/parsers/number.test.ts [hunk 1]
```diff
@@ -116,7 +118,15 @@ describe('number', () => {
...
       ({ type, value, hex }) => {
         expect(
-          bytesToHex(number.encode({ type, value, buffer: new Uint8Array() })),
+          bytesToHex(
+            number.encode({
+              type,
+              value,
+              buffer: new Uint8Array(),
+              packed: false,
+              tight: false,
```
[the remark this answers is not in the document]
Comment (FrederikBolding) [reply src/parsers/number.test.ts:127]: I mean, do we need fixtures for each parser or is it enough to do testing on the full fixtures?

File: src/packer.ts [hunk 2]
```diff
@@ -111,9 +158,17 @@ export const pack = <Type extends readonly string[]>(
       const parser = getParser(type);
       const value = values[index];
 
-      if (!isDynamicParser(parser, type)) {
+      // If packed mode is enabled, we can skip the dynamic check, as all
+      // values are encoded in the static buffer.
+      if (packed || arrayPacked || !isDynamicParser(parser, type)) {
```
Comment (FrederikBolding) [inline src/packer.ts:163]: Why is the extra `arrayPacked` necessary? 🤔

File: src/parsers/array.ts [hunk 3]
```diff
@@ -124,18 +145,34 @@ export const array: Parser<unknown[]> = {
...
+    }
+
+    if (packed) {
+      // For packed encoding, we don't need to encode the length of the array,
+      // so we can just encode the values.
+      return pack({
+        types: new Array(value.length).fill(arrayType),
+        values: value,
+        byteArray: buffer,
+        packed: fixedBytes.isType(arrayType) && tight,
```
Comment (FrederikBolding) [inline src/parsers/array.ts:160]: Not sure I understand the logic here 🤔

File: src/parsers/array.ts [hunk 4]
```diff
@@ -106,11 +107,31 @@ export const array: Parser<unknown[]> = {
...
+
+    // Tightly pack `T[]` where `T` is a dynamic type. This is not supported in
+    // Solidity, but is commonly used in the Ethereum ecosystem.
+    if (packed && isDynamicParser(getParser(arrayType), arrayType)) {
+      return pack({
+        types: new Array(value.length).fill(arrayType),
+        values: value,
+        byteArray: buffer,
+        packed,
+        arrayPacked: true,
```
Comment (FrederikBolding) [inline src/parsers/array.ts:131]: Missing `tight`?

File: src/parsers/parser.ts [hunk 5]
```diff
@@ -15,6 +15,16 @@ export type EncodeArgs<Value> = {
...
+
+  /**
+   * Whether to use the non-standard packed mode.
+   */
+  packed: boolean;
+
+  /**
+   * Whether to use tight packing mode. Only applicable when `packed` is true.
+   */
+  tight: boolean;
```
Comment (FrederikBolding) [inline src/parsers/parser.ts:27]: We pass this around, but it is not used? 🤔

File: src/abi.ts [hunk 6]
```diff
@@ -114,6 +121,41 @@ export const encodeSingle = <Type extends string>(
   return encode([type], [value]);
 };
 
+/**
+ * Encode the data with the provided types. The types must be valid Solidity
+ * ABI types. This is the same as {@link encode}, except that the values are
+ * encoded in the non-standard packed mode.
+ *
+ * Note that these values cannot be decoded with {@link decode} or Solidity's
```
[the remark this answers is not in the document]
Comment (Mrtenz) [reply src/abi.ts:129]: They can't, and that's not really the point. They're commonly used for creating signatures.

File: src/parsers/number.test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Mrtenz) [reply src/parsers/number.test.ts:127]: I've added some tests for the edge cases at least. Everything else should be tested quite well through the test vectors. But we can definitely improve on that in the future.

File: src/parsers/fixed-bytes.ts [hunk 7]
```diff
@@ -72,6 +73,10 @@ export const fixedBytes: Parser<BytesLike, Uint8Array> = {
       ),
     );
 
+    if (packed) {
+      return concatBytes([buffer, padEnd(bufferValue, length)]);
```
[the remark this answers is not in the document]
Comment (Mrtenz) [reply src/parsers/fixed-bytes.ts:77]: Yeah, values are always padded to their length, except for dynamic values (since those don't have a length).

`uint16` is always two bytes, `bytes8` is always eight bytes, and so on.

File: src/parsers/bytes.ts [hunk 8]
```diff
@@ -39,10 +39,16 @@ export const bytes: Parser<BytesLike, Uint8Array> = {
...
    * @param args.value - The bytes value to encode.
+   * @param args.packed - Whether to use packed encoding.
    * @returns The bytes with the encoded bytes value added to it.
    */
-  encode({ buffer, value }): Uint8Array {
+  encode({ buffer, value, packed }): Uint8Array {
     const bufferValue = createBytes(value);
+
+    if (packed) {
+      return concatBytes([buffer, bufferValue]);
```
[the remark this answers is not in the document]
Comment (Mrtenz) [reply src/parsers/bytes.ts:49]: The dynamic bytes type doesn't have a length.

File: src/parsers/number.ts [hunk 9]
```diff
@@ -145,20 +145,34 @@ export const number: Parser<NumberLike, bigint> = {
...
-  encode({ type, buffer, value }): Uint8Array {
+  encode({ type, buffer, value, packed }): Uint8Array {
     const bigIntValue = getBigInt(value);
 
     assertNumberLength(bigIntValue, type);
 
     if (isSigned(type)) {
+      if (packed) {
+        const length = getLength(type) / 8;
+        return concatBytes([buffer, signedBigIntToBytes(bigIntValue, length)]);
```
[the remark this answers is not in the document]
Comment (Mrtenz) [reply src/parsers/number.ts:159]: Yes, encoding to two's complement requires a certain length to be specified.

File: src/packer.ts (hunk 2 above)
Comment (Mrtenz) [reply src/packer.ts:163]: Array values are padded to 32 bytes, but the values are still encoded in place. I needed a way to skip adding the length, but still pad to 32 bytes.

File: src/parsers/array.ts (hunk 3 above)
Comment (Mrtenz) [reply src/parsers/array.ts:160]: This is a weird edge case in `ethereumjs-abi`. It disregards the required size (i.e., 32 bytes for arrays) and always encodes `bytesN` as their length.

https://github.com/ethereumjs/ethereumjs-abi/blob/ee3994657fa7a427238e6ba92a84d0b529bbcde0/lib/index.js#L479-L485

File: src/parsers/parser.ts (hunk 5 above)
Comment (Mrtenz) [reply src/parsers/parser.ts:27]: It's used in the `array` parser.

File: src/parsers/number.test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (FrederikBolding) [reply src/parsers/number.test.ts:127]: Please create an issue for that!

File: src/parsers/array.ts (hunk 3 above)
Comment (FrederikBolding) [reply src/parsers/array.ts:161]: I think we should add a comment that explains this!

File: src/parsers/array.ts [hunk 10]
```diff
@@ -106,11 +107,32 @@ export const array: Parser<unknown[]> = {
...
+   * @param args.packed - Whether to use non-standard packed encoding.
+   * @param args.tight - Whether to use non-standard tight encoding.
    * @returns The bytes with the encoded array added to it.
    */
-  encode({ type, buffer, value }): Uint8Array {
+  encode({ type, buffer, value, packed, tight }): Uint8Array {
     const [arrayType, fixedLength] = getArrayType(type);
 
+    // Packed encoding does not support nested arrays.
+    assert(
```
Comment (FrederikBolding) [inline src/parsers/array.ts:118]: Please test this error case

File: src/parsers/bytes.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (FrederikBolding) [reply src/parsers/bytes.ts:51]: For packed you mean? Because it looks like we encode it below for "normal mode"

File: src/packer.ts [hunk 11]
```diff
@@ -83,21 +83,68 @@ type PackState = {
...
+   * When true, the packed mode will not add any padding bytes. This matches
+   * the packing behaviour of `ethereumjs-abi`, but is not standard.
+   */
+  tight?: boolean | undefined;
+
+  /**
+   * Whether to use the non-standard packed mode in "array" mode. This is
+   * normally only used by the {@link array} parser.
+   */
+  arrayPacked?: boolean | undefined;
```
Comment (FrederikBolding) [inline src/packer.ts:113]: I wonder if this should be exposed to the user as it is sort of an implementation detail, what do you think?

File: src/packer.ts (hunk 2 above)
Comment (FrederikBolding) [reply src/packer.ts:163]: Hmm, I wonder if there is a better way 🤔

File: src/packer.ts (hunk 11 above)
Comment (Mrtenz) [reply src/packer.ts:113]: It's not, since `packer.ts` isn't exported directly. You have to call it through `encodePacked`, which does not have an option for this.

File: src/parsers/bytes.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (Mrtenz) [reply src/parsers/bytes.ts:51]: Oh, right. For normal encoding we encode the length of a dynamic byte array (and other dynamic values, like strings and arrays). It's required in order to decode the value. In packed mode we don't encode it.

[the remark this answers is not in the document]
Comment (FrederikBolding) [reply src/parsers/bytes.ts:51]: Perhaps expand the comment to mention that?

File: src/packer.ts (hunk 2 above)
Comment (FrederikBolding) [reply src/packer.ts:163]: Perhaps open an issue to revisit this?

File: src/packer.ts (hunk 11 above)
Comment (FrederikBolding) [reply src/packer.ts:113]: Ah, great!

File: src/parsers/array.ts [hunk 12]
```diff
@@ -124,18 +146,38 @@ export const array: Parser<unknown[]> = {
         type: getTupleType(arrayType, fixedLength),
         buffer,
         value,
+        // In "tight" mode, we don't pad the values to 32 bytes if the value is
```
Comment (FrederikBolding) [inline src/parsers/array.ts:149]: Can we add this comment in both places?

File: src/parsers/number.test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Mrtenz) [reply src/parsers/number.test.ts:127]: #18

File: src/packer.ts (hunk 2 above)
Comment (Mrtenz) [reply src/packer.ts:163]: #19

**Review by FrederikBolding — APPROVED** [review]

**Review by mannyad — APPROVED** [review]
Not bad

**Review by mannyad — APPROVED** [review]
That nice
